libtransact/src/scheduler/serial/core.rs

```diff
@@ -156,92 +170,150 @@ impl SchedulerCore {
             return Ok(());
         }
 
-        let mut shared = self
-            .shared_lock
-            .lock()
-            .expect("scheduler shared lock is poisoned");
+        if self.current_batch.is_none() {
+            let mut shared = self
+                .shared_lock
+                .lock()
+                .expect("scheduler shared lock is poisoned");
+
+            if let Some(unscheduled_batch) = shared.pop_unscheduled_batch() {
+                self.txn_queue = unscheduled_batch.batch().transactions().to_vec();
+                self.current_batch = Some(unscheduled_batch);
+            } else {
+                return Ok(());
+            }
+        }
 
-        if let Some(pair) = shared.pop_unscheduled_batch() {
-            if pair.batch().transactions().len() != 1 {
-                unimplemented!("can't handle more than one transaction per batch");
+        let transaction = self
+            .txn_queue
+            .pop()
+            .expect("There are no transactions left in the queue for the current batch");
+        let transaction_id = transaction.header_signature().into();
+        let transaction_pair = match transaction.into_pair() {
+            Ok(pair) => pair,
+            Err(err) => {
+                self.invalidate_current_batch(InvalidTransactionResult {
+                    transaction_id,
+                    error_message: format!("ill-formed transaction: {}", err),
+                    error_data: vec![],
+                });
+                self.send_batch_result();
+                return Ok(());
             }
+        };
 
-            let transaction = match pair.batch().transactions()[0].clone().into_pair() {
-                Ok(transaction) => transaction,
-                Err(err) => {
-                    // An error can occur here if the header of the
-                    // transaction cannot be deserialized. If this occurs,
-                    // we panic as handling this is not currently implemented.
-                    unimplemented!("cannot handle ill-formed transaction: {}", err)
-                }
-            };
-
-            let context_id = match self.previous_context {
-                Some(previous_context_id) => self
-                    .context_lifecycle
-                    .create_context(&[previous_context_id], &self.state_id),
-                None => self.context_lifecycle.create_context(&[], &self.state_id),
-            };
-
-            self.current_batch = Some(pair);
-            self.execution_tx
-                .send(ExecutionTask::new(transaction, context_id))?
-        }
+        let context_id = match self.previous_context {
+            Some(previous_context_id) => self
+                .context_lifecycle
+                .create_context(&[previous_context_id], &self.state_id),
+            None => self.context_lifecycle.create_context(&[], &self.state_id),
+        };
+
+        self.current_txn = Some(transaction_pair.clone());
```

Comment:
We might want to avoid this clone for performance reasons, but without it we would have to track via indexes instead of the current/queue approach.

Reply:
Really all we need is the transaction ID; that would still require a clone, but it would be less expensive. Otherwise the index approach would be necessary.

Reply:
I updated `current_txn` to store just the transaction ID rather than the whole transaction pair.